chain/chain/src/resharding/manager.rs

```diff
@@ -250,6 +272,99 @@ impl ReshardingManager {
         Ok(())
     }
 
+    fn get_child_congestion_info(
+        tries: &ShardTries,
+        parent_shard_layout: &ShardLayout,
+        parent_shard_uid: ShardUId,
+        parent_chunk_extra: &Arc<ChunkExtra>,
+        child_shard_layout: &ShardLayout,
+        child_shard_uid: ShardUId,
+        retain_mode: RetainMode,
+    ) -> Result<CongestionInfo, Error> {
+        let parent_congestion_info =
+            parent_chunk_extra.congestion_info().expect("The congestion info must exist!");
+
+        // Get the congestion info based on the parent shard.
+        let parent_state_root = *parent_chunk_extra.state_root();
+        let parent_trie = tries.get_trie_for_shard(parent_shard_uid, parent_state_root);
+        let mut child_congestion_info = Self::get_child_congestion_info_not_finalized(
+            &parent_trie,
+            &parent_shard_layout,
+            parent_congestion_info,
+            retain_mode,
+        )?;
+
+        // Set the allowed shard based on the child shard.
+        Self::finalize_allowed_shard(
+            &child_shard_layout,
+            child_shard_uid,
+            &mut child_congestion_info,
+        )?;
+
+        Ok(child_congestion_info)
+    }
+
+    // Get the congestion info for the child shard. The congestion info can be
+    // inferred efficiently from the combination of the parent shard's
+    // congestion info and the receipt group metadata, that is available in the
+    // parent shard's trie.
+    pub fn get_child_congestion_info_not_finalized(
+        parent_trie: &dyn TrieAccess,
+        parent_shard_layout: &ShardLayout,
+        parent_congestion_info: CongestionInfo,
+        retain_mode: RetainMode,
+    ) -> Result<CongestionInfo, Error> {
+        // The left child contains all the delayed and buffered receipts from the
+        // parent so it should have identical congestion info.
+        if retain_mode == RetainMode::Left {
+            return Ok(parent_congestion_info);
+        }
+
+        // The right child contains all the delayed receipts from the parent but it
+        // has no buffered receipts. It's info needs to be computed by subtracting
+        // the parent's buffered receipts from the parent's congestion info.
+        let mut congestion_info = parent_congestion_info;
+        for shard_id in parent_shard_layout.shard_ids() {
+            let receipt_groups = ReceiptGroupsQueue::load(parent_trie, shard_id)?;
+            let Some(receipt_groups) = receipt_groups else {
+                continue;
+            };
+
+            let bytes = receipt_groups.total_size();
+            let gas = receipt_groups.total_gas();
+            let gas = gas.try_into().expect("ReceiptGroup gas must fit in u64");
+
+            congestion_info
```

Comment:
do we not have to set `total_receipts_num` as well?

Reply:
Good sanity check! The `total_receipts_num` field is part of `ReceiptGroupsQueueData` which is resharded by copying to left child only. It is handled by the memtrie and flat storage split operations. MemTrie is already implemented [here](https://github.com/near/nearcore/blob/33afc1d35dffb4660d8c3423385540004704cc09/core/store/src/trie/ops/resharding.rs#L60-L68) and this PR adds the flat storage equivalent.